File: Morhun_transfusion_task_infinity.pl
% Задаємо ємність посудин (четверта посудина — нескінченний кран/стік, позначаємо великою кількістю)
capacity([8, 3, 7, 1000]).

% Задаємо початковий стан і фінальний стан
start_stage([8, 0, 0, 0]).
goal_stage([3, 0, 5, _]).  % Четверта посудина може бути в будь-якому стані

% Операція переливання води з однієї посудини в іншу
pour([X, Y, Z, W], [X1, Y1, Z, W]) :- capacity([_, MaxY, _, _]), transfer(X, Y, X1, Y1, MaxY).
pour([X, Y, Z, W], [X1, Y, Z1, W]) :- capacity([_, _, MaxZ, _]), transfer(X, Z, X1, Z1, MaxZ).
pour([X, Y, Z, W], [X, Y1, Z1, W]) :- capacity([_, _, MaxZ, _]), transfer(Y, Z, Y1, Z1, MaxZ).
pour([X, Y, Z, W], [X1, Y1, Z, W]) :- capacity([MaxX, _, _, _]), transfer(Y, X, Y1, X1, MaxX).
pour([X, Y, Z, W], [X1, Y, Z1, W]) :- capacity([MaxX, _, _, _]), transfer(Z, X, Z1, X1, MaxX).
pour([X, Y, Z, W], [X, Y1, Z1, W]) :- capacity([_, MaxY, _, _]), transfer(Z, Y, Z1, Y1, MaxY).

% Додаємо переливання з крана / в стік
pour([X, Y, Z, W], [X1, Y, Z, W1]) :- capacity([_, _, _, MaxW]), transfer(X, W, X1, W1, MaxW). % Виливаємо в стік
pour([X, Y, Z, W], [X, Y1, Z, W1]) :- capacity([_, _, _, MaxW]), transfer(Y, W, Y1, W1, MaxW).
pour([X, Y, Z, W], [X, Y, Z1, W1]) :- capacity([_, _, _, MaxW]), transfer(Z, W, Z1, W1, MaxW).
pour([X, Y, Z, W], [X1, Y, Z, W1]) :- capacity([_, _, _, MaxW]), transfer(W, X, W1, X1, MaxW). % Наповнюємо з крана
pour([X, Y, Z, W], [X, Y1, Z, W1]) :- capacity([_, _, _, MaxW]), transfer(W, Y, W1, Y1, MaxW).
pour([X, Y, Z, W], [X, Y, Z1, W1]) :- capacity([_, _, _, MaxW]), transfer(W, Z, W1, Z1, MaxW).

% Функція переливу з однієї посудини в іншу
transfer(A, B, A1, B1, MaxB) :-  
    Total is A + B,  
    (Total =< MaxB -> A1 = 0, B1 = Total ; A1 is Total - MaxB, B1 = MaxB).

% Основний пошук шляху
path(P) :-
    start_stage(Start),
    goal_stage(Goal),
    bfs_algorithm([[Start]], Goal, RevPath),
    reverse(RevPath, P).

% Алгоритм пошуку в ширину
bfs_algorithm([[Goal | Rest] | _], Goal, [Goal | Rest]).
bfs_algorithm([[State | Rest] | Queue], Goal, P) :-
    findall([Next, State | Rest],
            (pour(State, Next), \+ member(Next, [State | Rest])),
            NewPaths),
    append(Queue, NewPaths, NewQueue),
    bfs_algorithm(NewQueue, Goal, P).

File: Morhun_transfusion_task_infinity.hs
import Data.List (nub)

type State = (Int, Int, Int, Int)

capacity :: (Int, Int, Int, Int)
capacity = (9, 4, 7, 1000)  -- 4-та посудина - нескінченне джерело/стік

startStage :: State
startStage = (9, 4, 0, 0)

goalStage :: State -> Bool
goalStage (6, 3, 4, _) = True
goalStage _ = False

transfer :: Int -> Int -> Int -> (Int, Int)
transfer a b maxB =
    let total = a + b
    in if total <= maxB then (0, total) else (total - maxB, maxB)

pour :: State -> [State]
pour (x, y, z, w) =
    let (maxX, maxY, maxZ, maxW) = capacity
    in nub $ filter (/= (x, y, z, w))  -- Виключаємо поточний стан
        [ let (x1, y1) = transfer x y maxY in (x1, y1, z, w)
        , let (x1, z1) = transfer x z maxZ in (x1, y, z1, w)
        , let (y1, z1) = transfer y z maxZ in (x, y1, z1, w)
        , let (y1, x1) = transfer y x maxX in (x1, y1, z, w)
        , let (z1, x1) = transfer z x maxX in (x1, y, z1, w)
        , let (z1, y1) = transfer z y maxY in (x, y1, z1, w)
        , let (x1, w1) = transfer x w maxW in (x1, y, z, w1)
        , let (y1, w1) = transfer y w maxW in (x, y1, z, w1)
        , let (z1, w1) = transfer z w maxW in (x, y, z1, w1)
        , let (w1, x1) = transfer w x maxX in (x1, y, z, w1)
        , let (w1, y1) = transfer w y maxY in (x, y1, z, w1)
        , let (w1, z1) = transfer w z maxZ in (x, y, z1, w1)
        ]

bfs :: [[State]] -> Maybe [State]
bfs [] = Nothing
bfs (path@(current:_) : queue)
    | goalStage current = Just (reverse path)
    | otherwise =
        let newPaths = [next:path | next <- pour current, notElem next path]
        in bfs (queue ++ newPaths)

solve :: Maybe [State]
solve = bfs [[startStage]]

main :: IO ()
main = case solve of
    Just path -> mapM_ print path
    Nothing   -> putStrLn "No solution found"
